### PR TITLE
[#685] Add support for long file paths.

### LINKED ISSUE #685: Add support for long file names
We need to enable the tests on long paths

and all file operations on windows should add the `\\?\` to the path .

For local shares, we might need to add `\\?\UNC\` 

```
>>> pa[4:]
u'C:\\Users\\chevah\\Downloads\\very-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very\\very-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very'
>>> os.listdir(pa[4:])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
WindowsError: [Error 3] The system cannot find the path specified: u'C:\\Users\\chevah\\Downloads\\very-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very\\very-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very\\*.*'
>>> os.listdir(pa)
[u'New Text Document.txt']
>>> pa
u'\\\\?\\C:\\Users\\chevah\\Downloads\\very-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very\\very-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very-very-longvery-long-name-bug-very'
```

### PR DESCRIPTION
**Pull request by adiroiban**
Scope
=====

Fixes #685

This tries to fix filesystem to handle long  file paths.

Changes
=======

Enable CI with temp on a long path and fix the implementation to  get the  tests green.

How to try and test the changes
===============================

reviewers: @danuker 

just a code review her.

We will do manual tests in chevh/server

### CODE REVIEW

File: src/chevah_compat/tests/normal/testing/test_testcase.py [hunk 1]
```diff
@@ -769,7 +769,12 @@ def test_cleanup_test_segments_link(self):
         When self.test_segments is defined it will be automatically
         removed, even when it is a symbolic link.
         """
-        _, self.test_segments = mk.fs.makePathInTemp()
+        # FIXME:
```
Comment (adiroiban) [inline src/chevah_compat/tests/normal/testing/test_testcase.py:772]: to be fixed before merge

**Comment by adiroiban** [comment]
For now this is only py2 code.

Once all is ok, the plan is to merge this into master and then merge the changes into #676

needs-review

File: src/chevah_compat/nt_filesystem.py [hunk 2]
```diff
@@ -158,7 +158,13 @@ def installation_segments(self):
...
         '''Return the encoded representation of the path, use in the lower
         lever API for accessing the filesystem.'''
-        return path
+        if path.startswith(u'\\\\?\\'):
+            # This might be already and encoded long path.
+            return path
+
+        if len(path) < 250:
+            return path
+        return u'\\\\?\\' + path
```
Comment (danuker) [inline src/chevah_compat/nt_filesystem.py:167]: Is this the reason for this PR?
Maybe add a link to this long path format being documented:

```suggestion
        # An extended-length path, use the "\\?\" prefix.
        # https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation
        return u'\\\\?\\' + path
```

File: src/chevah_compat/posix_filesystem.py [hunk 3]
```diff
@@ -749,6 +753,10 @@ def _dirEntryToFileAttributes(self, entry):
             # On Windows, scandir gets float precision while
             # getAttributes only integer.
             modified = int(modified)
+            # On Windows, path might have long names for local drives.
+            # For compat, we keep the simple format.
+            if path.startswith('\\\\?\\') and path[5] == ':':
+                path = path[4:]
```
Comment (danuker) [inline src/chevah_compat/posix_filesystem.py:759]: If it's a known extended-length path, why remove the `\\?\` prefix?

**Review by danuker — APPROVED** [review]
Looks merge-able to me. But I would add a link to the extended-length path format.

File: src/chevah_compat/posix_filesystem.py (hunk 3 above)
Comment (adiroiban) [reply src/chevah_compat/posix_filesystem.py:759]: the high level API should just use the normal path format.
Compat will handle the low-level conversion at the low-level.

Here we are exposing a low-level path format to the high-level user.
So we normalize/clean it.

In this way, you can be sure that when you do `result = fs.iterate('c:\very-long-folder\')` the resulting member will have paths like `c:\very-long-folder\some-name.txt` and not `\\?\c:\very-long-folder\some-name.txt`

Comment (danuker) [reply src/chevah_compat/posix_filesystem.py:762]: Yes! Makes sense. This path goes towards the user side for `_dirEntryToFileAttributes`. Thank you.

File: src/chevah_compat/testing/filesystem.py [hunk 4]
```diff
@@ -28,7 +28,12 @@ def __init__(self, avatar=None):
         Create an unique temp folder.
         """
         super(LocalTestFilesystem, self).__init__(avatar=avatar)
-        self._temp_uuid = u'%s%s' % (uuid.uuid4(), TEST_NAME_MARKER)
+        self._temp_uuid = u'%s%s%s' % (
+            'long-name-' * 25, uuid.uuid4(), TEST_NAME_MARKER)
+        # Make sure we keep the directory below 255.
+        # This is the limit for a single filename in Windows and ext4.
+        # Any file created inside will have a longer path.
+        self._temp_uuid = self._temp_uuid[-254:]
```
[the remark this answers is not in the document]
Comment (adiroiban) [reply src/chevah_compat/testing/filesystem.py:36]: I think that this is just Windows File Explorer GUI... and the strange thing is only about sorting.

i think that we are fine.

**Comment by adiroiban** [comment]
Many thanks @danuker for the review.

I am working  in parallel to get the chevah/server tests working and I might need to do minor changes before the final merge.

but I think that 99% is ok.

**Comment by adiroiban** [comment]
I am going to merge this as the chevah/server tests look green.

I will follow with the patch for py3 version.

**Pull request merged**